File: admin/src/theme/themeController.ts
import {
  decodeThemeChoice,
  type ResolvedTheme,
  type ThemeChoice,
  type ThemePreferenceStore,
} from './themePreference.ts'

export class ThemeController {
  mode: ThemeChoice
  readonly #store: ThemePreferenceStore

  constructor(store: ThemePreferenceStore) {
    this.#store = store
    this.mode = decodeThemeChoice(store.read())
  }

  load(): void {
    this.mode = decodeThemeChoice(this.#store.read())
  }

  toggle(resolved: ResolvedTheme): void {
    this.mode = resolved === 'dark' ? 'light' : 'dark'
    try {
      this.#store.write(this.mode)
    } catch {
      // Keep the session choice when browser storage is unavailable.
    }
  }
}

File: app/lib/src/theme/theme_controller.dart
import 'package:flutter/material.dart';

import 'theme_preference_store.dart';

class ThemeController extends ChangeNotifier {
  ThemeController({
    ThemePreferenceStore? store,
    ThemeMode initial = ThemeMode.system,
  }) : _store = store,
       _mode = initial;

  final ThemePreferenceStore? _store;
  ThemeMode _mode;

  ThemeMode get mode => _mode;

  Future<void> load() async {
    final stored = await _store?.read();
    _mode = _decode(stored);
    notifyListeners();
  }

  Future<void> toggle(Brightness resolvedBrightness) async {
    _mode = resolvedBrightness == Brightness.dark
        ? ThemeMode.light
        : ThemeMode.dark;
    notifyListeners();
    await _store?.write(_mode.name);
  }

  static ThemeMode _decode(String? stored) {
    return switch (stored) {
      'light' => ThemeMode.light,
      'dark' => ThemeMode.dark,
      _ => ThemeMode.system,
    };
  }
}
